[Feat] 애플 로그인시 신규유저인지 판단하는 로직과 소셜 로그인 유저의 회원가입하는 로직을 추가하였습니다.

snughnu opened the pull request:
# 배경
<img width="298" alt="스크린샷 2025-03-02 21 51 16" src="https://github.com/user-attachments/assets/8126caa2-3e54-435d-9d9e-9ad933023ad0" />

- 애플 로그인 버튼을 터치하였을 때의 동작을 구현했습니다.
- 관련 이슈 및 PR: #16 #93 #77 

# 작업 내용
## 테스트 영상
https://github.com/user-attachments/assets/53aade45-a76c-41ea-87af-60a4811635bc
- 애플 로그인 버튼을 터치 후, 로그인 성공 시 "로그인 성공",
- idToken이 키체인에 성공적으로 저장되면 "idToken 저장 성공" 이 로그에 출력됩니다.
- 현재 idToken을 이용해 신규 유저인지 판별하는 로직이 작성되지 않아, UI에서는 "애플 로그인 실패" 라고 표시됩니다.

## 카카오 로그인 후, 신규유저일 때 회원가입하는 로직
- 카카오 로그인에서 신규유저일 때 회원가입하는 로직을 추가 및 수정하였습니다.

## 구조 및 `AppleLoginManager` 사용
- 기본적인 구조는 MVVM + Clean 아키텍처이며, `SocialLoginRepository`에서 `AppleLoginManager`를 사용하여 메서드를 실행하는 구조입니다.

### `AppleLoginManager` 구성
- `loginWithApple(completion:)`
  - 애플 로그인 요청을 수행하고, 성공 시 idToken을 반환합니다.
- `authorizationController(controller:didCompleteWithAuthorization:)`
  - 애플 로그인 인증이 완료되었을 때, idToken을 추출하여 반환합니다.
- `presentationAnchor(for:)`
  - 로그인 팝업을 표시할 기준이 되는 UIWindow를 제공합니다.

## idToken을 이용한 신규 유저인지 판별 로직
- ~~해당 API가 아직 제공되지 않아서 현재는 기본적인 구조만 작성해두었습니다.~~
- ~~API가 완성되면 해당 로직을 추가 및 수정하고 PR을 다시 올릴 예정입니다.~~
- idToken을 이용하여 신규 유저인지 판별하는 로직을 추가하였습니다.

## idToken과 LoginType을 이용하여 회원가입 로직 추가
- 키체인에 idToken과 LoginType을 저장하는 로직을 추가하였고,
이를 사용하여 회원가입화면2로 갔을 때 카카오, 애플, 자체 회원가입의 분기처리를 하였습니다.

# 테스트 방법
```
// MARK: - for test
deleteTokensForTest()
func deleteTokensForTest() {
    let deleteQueries: [[String: Any]] = [
        [kSecClass as String: kSecClassGenericPassword, kSecAttrAccount as String: "Popcorn", kSecAttrService as String: "accessToken"],
        [kSecClass as String: kSecClassGenericPassword, kSecAttrAccount as String: "Popcorn", kSecAttrService as String: "refreshToken"],
        [kSecClass as String: kSecClassGenericPassword, kSecAttrAccount as String: "Popcorn", kSecAttrService as String: "accessExpiredAt"],
        [kSecClass as String: kSecClassGenericPassword, kSecAttrAccount as String: "Popcorn", kSecAttrService as String: "refreshExpiredAt"],
        [kSecClass as String: kSecClassGenericPassword, kSecAttrAccount as String: "Popcorn", kSecAttrService as String: "loginType"]
    ]

    for query in deleteQueries {
        let status = SecItemDelete(query as CFDictionary)
        if status == errSecSuccess {
            print("✅ \(query[kSecAttrService as String]!) 삭제 성공")
        } else {
            print("❌ \(query[kSecAttrService as String]!) 삭제 실패: \(status)")
        }
    }
    print("🚨 테스트: 모든 토큰 및 만료 날짜 정보가 삭제되었습니다.")
}
```
로 Token을 삭제 후, 시뮬레이터에서 iCloud 로그인 후, 애플 로그인 버튼을 눌러 테스트할 수 있습니다.

# 리뷰 노트
- ~~idToken을 이용한 신규 유저 판단 API가 완성되면, 해당 로직을 추가한 후 다시 PR을 올리겠습니다!~~
  - 코드의 변동이 적고 현재 PR에 맞는 내용이기에 새로운 PR이 아닌 기존 PR에 추가하였습니다.
- PR분리
  - 원래는 카카오 로그인 후 회원가입을 수행하는 로직을 별도의 PR로 관리하려 했으나, 충돌 해결 과정에서 함께 커밋되었습니다.
  해당 부분을 리뷰하실 때 참고 부탁드립니다!

snughnu committed:
setup: 인증서 추가

snughnu committed:
feat: DIContainer 수정

- DIContainer에서의 메서드를 없애고 앱델에서 등록

snughnu committed:
feat: Apple로그인 코드 베이스 작성

snughnu committed:
feat: 카카오 회원가입 로직 수정

snughnu committed:
refactor: 관심사 매핑 로직 수정

snughnu committed:
feat: 애플 회원가입 코드 베이스 작성

- 회원가입response DTO rename

snughnu committed:
feat: AppleLoginManagerProtocol 생성

snughnu committed:
feat: AppleLoginManager 추가 및 saveIdToken 로직 추가

MinwooJe commented on Popcorn-iOS/Source/Domain/UseCases/LoginUseCase/SocialLoginUseCase.swift:
```diff
@@ -55,6 +56,36 @@ final class SocialLoginUseCase: SocialLoginUseCaseProtocol {
+        switch result {
+        case .success(let idToken):
+            socialLoginRepository.fetchNewAppleUserResult(idToken: idToken.idToken) { [weak self] result in
+                guard let self = self else { return }
```
p4)
별건 아니지만 Swift 5.7에서 도입 된 `shorthand optional binding`을 사용하면
`guard let self else { return }` 처럼 사용 가능합니다!

> https://zeddios.tistory.com/1359

snughnu replied:
헐 맞네요, 나중에 바꿔보도록 하겠습니다!

MinwooJe commented on Popcorn-iOS/Source/Data/Keychain/KeychainManager.swift:
```diff
@@ -84,3 +88,51 @@ final class KeychainManager: KeychainManagerProtocol {
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrAccount as String: "idToken"
+        ]
+
+        let attributes: [String: Any] = [
+            kSecValueData as String: data,
+            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock
+        ]
+
+        let status = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
+        if status == errSecItemNotFound {
+            let addQuery = query.merging(attributes) { _, new in new }
+            return SecItemAdd(addQuery as CFDictionary, nil)
+        }
+        return status
+    }
+
+    func fetchIdToken() -> String? {
+        let query: [String: Any] = [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrAccount as String: "idToken",
+            kSecReturnData as String: true,
+            kSecMatchLimit as String: kSecMatchLimitOne
+        ]
+
+        guard let data = fetchItem(with: query) else { return nil }
+        return String(data: data, encoding: .utf8)
+    }
+
+    // MARK: - Delete IdToken
+    @discardableResult
+    func deleteIdToken() -> OSStatus {
+        let query: [String: Any] = [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrAccount as String: "idToken"
+        ]
+
+        return deleteItem(with: query)
+    }
+}
```
p2)
KeychainManager는 토큰 뿐만이 아니라 사용자의 민감한 정보를 관리하는 역할입니다.
따라서 KeychainManger는 단순 CRUD(add, fetch, update, delete) 메서드만 가지는 것이 적절 할 것 같습니다!
그러므로 loadSignupData()와 토큰과 관련된 디테일한 로직은 KeychainManeger가 아닌 Repository 쪽에 구현하는 것이 객체지향적으로 적절한 설계라고 생각됩니다!

snughnu replied:
넵! KeychainManager에 CRUD메서드만 남기고 이를 이용하는 로직으로 변경했습니다!

MinwooJe commented on Popcorn-iOS/Source/Data/Network/AppleLoginManager/AppleLoginManager.swift:
```diff
@@ -10,4 +10,53 @@ import AuthenticationServices
 }
+
+final class AppleLoginManager: NSObject, AppleLoginManagerProtocol {
+    private var completion: ((Result<IdToken, Error>) -> Void)?
```
p3)
질문입니다!
`completion`을 프로퍼티로 선언한 이유가 어떤걸까요?

snughnu replied:
`authorizationController`에서 `completion`을 사용하는 구조이므로, `loginWithApple(completion:)` 내부에서 `completion`을 지역 변수로 선언하면 로그인 과정이 끝날 때까지 유지되지 않습니다.
애플 로그인은 비동기적으로 진행되고 최종 결과는 `didCompleteWithAuthorization` 또는 `didCompleteWithError`에서 전달되므로 `completion`을 프로퍼티로 저장하여 로그인 완료 후에도 실행할 수 있도록 했습니다!

MinwooJe replied:
아하 그렇군요! 감사합니다~~

MinwooJe reviewed (commented):
수고 많으셨습니다!
PR 본문이 잘 작성되어있어서 리뷰하기 좋네요!!
리뷰 노트도 잘 작성해주셔서 좋았습니다~~

snughnu committed:
refactor: 키체인에 crud만 남기기 위한 loadSignupData메서드 삭제

snughnu committed:
refactor: fetchIdToken메서드와 fetchDeleteIdTokenResult메서드 삭제

snughnu committed:
feat: saveIdToken메서드 삭제, updateItem 메서드를 사용

snughnu committed:
feat: 애플로그인, 회원가입 APIConstant 추가

snughnu committed:
feat: 애플 회원가입 로직 추가

- saveIdTokenAndLoginType메서드를 추가하여 반복 최소화
- SignUpRepository에 fetchAppleSignUpResult 와 fetchLoginType메서드 추가
- SignUpUseCase에서 로그인타입에 따른 분기 회원가입 로직 추가

MinwooJe commented on Popcorn-iOS/Source/Domain/UseCases/SignUpUseCase/SignUpUseCase.swift:
```diff
@@ -173,39 +171,52 @@ extension SignUpUseCase {
 
-        if let idToken = signUpRepository.fetchIdToken() {
-            let kakaoSignUpData = SocialSignUpRequestDTO(
+        if let idToken = signUpRepository.fetchIdToken(),
+           let loginType = signUpRepository.fetchLoginType() {
+
+            let socialSignUpData = SocialSignUpRequestDTO(
```
p4)
질문입니다!
회원가입 중인데, idToken과 loginType이 왜 필요한지 궁금합니다!!

snughnu replied:
`idToken`은 다음과 같은 역할을 합니다!
- 서버에서 소셜 로그인(Kakao, Apple) 사용자가 신규 유저인지 아닌지 확인하기 위해 필요함.
- 서버는 `idToken`을 db에서 검증하여:
  - 이미 가입된 유저라면 메인화면으로 이동
  - 신규 유저라면 회원가입으로 이동
    - 이때 서버는 `idToken`을 기반으로 회원가입한 내용을 db에 저장합니다.
    - 그래서 회원가입에 `idToken`이 필요합니다!

`loginType`은 다음과 같은 역할을 합니다!
- 회원가입 화면(두 번째 단계)에서 사용자가 어떤 방식으로 로그인했는지(카카오, 애플, 자체 회원가입)를 알아야 함.
- 회원가입 버튼을 눌렀을 때 `loginType`에 따라 회원가입 요청을 다르게 보낼 수 있도록 분기 처리해야 함.
  - 카카오 로그인한 경우 → `fetchKakaoSignUpResult()`
  - 애플 로그인한 경우 → `fetchAppleSignUpResult()`
  - 자체 회원가입한 경우 → `fetchSignUpResult()`
- 만약 `loginType`을 저장하지 않는다면 회원가입 시 어떤 API를 호출해야 할지 알 수 없기 때문에 `loginType`이 필요합니다!

MinwooJe replied:
아하 그렇군요 상세한 답변 감사드립니다!!!!
